x-pack/plugins/spaces/public/share_saved_objects_to_space/components/share_to_space_flyout_internal.tsx

```diff
@@ -85,27 +85,48 @@ function createDefaultChangeSpacesHandler(
     spacesToRemove: string[]
   ) => {
     const { title } = object;
-    const objectsToUpdate = objects.map(({ type, id }) => ({ type, id })); // only use 'type' and 'id' fields
+    const objectsToUpdate: Array<{ type: string; id: string }> = objects.map(({ type, id }) => ({
+      type,
+      id,
+    })); // only use 'type' and 'id' fields
     const relativesCount = objects.length - 1;
     const toastTitle = i18n.translate('xpack.spaces.shareToSpace.shareSuccessTitle', {
       values: { objectNoun: object.noun },
       defaultMessage: 'Updated {objectNoun}',
       description: `Object noun can be plural or singular, examples: "Updated objects", "Updated job"`,
     });
-    await spacesManager.updateSavedObjectsSpaces(objectsToUpdate, spacesToAdd, spacesToRemove);
+
+    // If removing spaces and there are referenced objects ("related objects" in UI),
+    // only remove spaces from the target object.
+    if (spacesToRemove.length > 0 && objectsToUpdate.length > 1) {
+      const indexOfTarget = objectsToUpdate.findIndex((element) => element.id === object.id);
+      if (indexOfTarget >= 0) {
+        objectsToUpdate.splice(indexOfTarget, 1);
+      }
+
+      await spacesManager.updateSavedObjectsSpaces(
+        [{ type: object.type, id: object.id }],
+        spacesToAdd,
+        spacesToRemove
+      );
+      if (spacesToAdd.length > 0) {
+        await spacesManager.updateSavedObjectsSpaces(objectsToUpdate, spacesToAdd, []);
+      }
```

Comment:
We now have two operations instead of one for the case where there are referenced objects. What would happen when one of these fails? Could we combine them into one? 

You're also calling them sequentially, could we run them in parallel to avoid unnecessary waiting?

Reply:
Good point. Larry and I talked about this, and I meant to address it before opening up for review.

We cannot really combine them without introducing additional changes and complexity to the HTTP API. The long term plan is to move this logic sever-side, which will better facilitate a single transaction solution. For now, I think the safest approach would be to perform the operation on the target object first. Only if that succeeds, proceed to the potential secondary operation on the related objects. We probably should also catch any errors so that we can re-throw with a more detailed description. I noticed that when this throws, it does not inform the user with anything useful. Does that sound like a good plan? Any suggestions?

Reply:
Agreed, changing the REST API is a bigger change than is necessary for this. 

As suggested I would still call both operations in parallel though until we've got a new API.

Reply:
> You're also calling them sequentially, could we run them in parallel to avoid unnecessary waiting?

As the UI only allows for a single object to be targeted, the overall count of objects for the operations will be relatively low. And if sharing the target object fails, sharing of its related objects probably shouldn't occur. The second operation also only ever occurs when adding and removing spaces at the same time. Because of this I am inclined to stage the second operation based on success of the first. However, I am open to running them in parallel if there is a good argument to do so.